###  Exercise 3: Load Data to Python!

Often, we store the data collected in files. xlsx and xls are two most common formats. To analyze data in the spreadsheet, the first thing we need to do is to load data tables from spreadsheet files to Python.

For the purpose of tutorial, I will provide you with a dataset to work with. You can click here to download the Excel file if you want to open it locally. This Excel file contains some information about airports in Alaska released by the Federal Aviation Association (FAA).

Pandas is an open source, BSD-licensed library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language. It provides convenient ways to work with data tables in Python which we will be utilizing in this guided project.


In [1]:
import pandas as pd

In [25]:
df_facilities = pd.read_excel("./Airport_Data.xlsx", sheet_name="Facilities")

#### To check what the dataset looks like, we can call the variable containing the data.

In [3]:
df_facilities

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
0      50009.*A        AIRPORT       'ADK     3/30/2017    AAL           NONE   
1     50016.1*A        AIRPORT       'AKK     3/30/2017    AAL           NONE   
2      50017.*A        AIRPORT       'Z13     3/30/2017    AAL           NONE   
3     50017.1*C  SEAPLANE BASE       'KKI     3/30/2017    AAL           NONE   
4      50020.*A        AIRPORT       'AKI     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
746  50920.12*C  SEAPLANE BASE       '2Y3     3/30/2017    AAL           NONE   
747    50920.*A        AIRPORT       'YAK     3/30/2017    AAL           NONE   
748   50925.1*A        AIRPORT       'A77     3/30/2017    AAL           NONE   
749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL           NONE   
750    50990.*A        AIRPORT       'L20     3/30/2017    AAL           NONE   

    State StateName             County CountyState  ... AirportPositionSource  \
0      AK    ALASKA     ALEUTIANS WEST          AK  ...      3RD PARTY SURVEY   
1      AK    ALASKA      KODIAK ISLAND          AK  ...                  NACO   
2      AK    ALASKA             BETHEL          AK  ...                 STATE   
3      AK    ALASKA             BETHEL          AK  ...                   NaN   
4      AK    ALASKA             BETHEL          AK  ...                 STATE   
..    ...       ...                ...         ...  ...                   ...   
746    AK    ALASKA    SKAGWAY-YAKUTAT          AK  ...               FAA-EST   
747    AK    ALASKA    SKAGWAY-YAKUTAT          AK  ...      3RD PARTY SURVEY   
748    AK    ALASKA          KUSKOKWIM          AK  ...                   NaN   
749    AK    ALASKA  KETCHIKAN GATEWAY          AK  ...                   NaN   
750    AK    ALASKA        UPPER YUKON          AK  ...               FAA-EST   

    AirportPositionSourceDate AirportElevationSource  \
0                    00:00:00       3RD PARTY SURVEY   
1                    00:00:00                   NACO   
2                    00:00:00                  STATE   
3                         NaN                    NaN   
4                    00:00:00                  STATE   
..                        ...                    ...   
746                  00:00:00                    NaN   
747                  00:00:00       3RD PARTY SURVEY   
748                       NaN                    NaN   
749                       NaN                    NaN   
750                  00:00:00                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
0                     00:00:00                   NaN              HGR   
1                     00:00:00                   NaN              NaN   
2                     00:00:00                   NaN              NaN   
3                          NaN                   NaN              NaN   
4                     00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
746                        NaN                   NaN              TIE   
747                   00:00:00                   NaN              HGR   
748                        NaN                   NaN              NaN   
749                        NaN                   NaN              NaN   
750                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
0           CARGO             Y           PADK           SS-SR  
1             NaN             Y           PAKH             NaN  
2           CARGO           Y-L            NaN         SEE RMK  
3             NaN             N            NaN             NaN  
4           CARGO           Y-L           PFAK           SS-SR  
..            ...           ...            ...             ...  
746         CARGO           NaN            NaN        

 #### As shown in the output of the above cell, this data sheet has 751 rows and 103 columns in total. Now you've successfully loaded the data from spreadsheet file into Python.

#### Exercise 4: Working with Data in Python
Now that you've loaded your data into Python, we can utilize the data science tool Pandas to explore the data and answer a lot of interesting questions!

#### Question 1: What are the types for aviation facilities in Alaska?
In this dataset, there is a variable called Type that records the type of a specific aviation facility. To answer this question, we would need to know what are the unique entries for this variable.

So naturally, we'll need to first extract this column from the entire dataset. The way we extract a column from a dataframe is similar to how we extract a column from a Python dictionary, except that we pass in the column (variable) name instead of key value:

In [5]:
df_facilities["Type"]

0            AIRPORT
1            AIRPORT
2            AIRPORT
3      SEAPLANE BASE
4            AIRPORT
           ...      
746    SEAPLANE BASE
747          AIRPORT
748          AIRPORT
749    SEAPLANE BASE
750          AIRPORT
Name: Type, Length: 751, dtype: object

#### The unique method of the returned Panda Series will give you the unique entires inside, and in this case, the answer to Question 1.

In [6]:
df_facilities["Type"].unique()

array(['AIRPORT', 'SEAPLANE BASE', 'HELIPORT'], dtype=object)

#### Question 2: Provide a table containing the information for all Seaplane Bases in Alaska
In this dataset, there is a variable called Type that records what is the type of a specific aviation facility. We want to filter out "Seaplane Base" on variable Type. Pandas provides a convenient way to perform element-wise comparison:

In [9]:
df_facilities["Type"] == "SEAPLANE BASE"

0      False
1      False
2      False
3       True
4      False
       ...  
746     True
747    False
748    False
749     True
750    False
Name: Type, Length: 751, dtype: bool

#### df_facilities[df_facilities["Type"] == "SEAPLANE BASE"] can be interpreted as:

Filter dataframe df_facilties conditioning on variable Type being strictly equal to "SEAPLANE BASE".

And the table shown above is the answer to Question 2.

#### Question 3: What are the 5 highest aviation facilities in Alaska?
In this dataset, the ARPElevation variable records the elevation of aviation facilities in Alaska in feet. What we will do is to sort the rows of this dataset in reverse based on the value of column ARPElevation. This can be done with sort_values method of Pandas dataframe:

In [10]:
df_facilities.sort_values(by = "ARPElevation", ascending = False)

SiteNumber           Type LocationID EffectiveDate Region DistrictOffice  \
584   50739.1*A        AIRPORT      'AA09     3/30/2017    AAL           NONE   
630  50738.20*A        AIRPORT      '93AK     3/30/2017    AAL           NONE   
265   50324.6*A        AIRPORT       '4Z5     3/30/2017    AAL           NONE   
582  50738.01*A        AIRPORT      '2AK7     3/30/2017    AAL           NONE   
353  50439.18*A        AIRPORT      'AK80     3/30/2017    AAL           NONE   
..          ...            ...        ...           ...    ...            ...   
183   50196.3*C  SEAPLANE BASE       'ELV     3/30/2017    AAL           NONE   
516    50680.*C  SEAPLANE BASE       'A23     3/30/2017    AAL           NONE   
363   50455.6*C  SEAPLANE BASE       '13Z     3/30/2017    AAL           NONE   
382    50489.*C  SEAPLANE BASE       'MTM     3/30/2017    AAL           NONE   
628   50773.5*C  SEAPLANE BASE      'AK62     3/30/2017    AAL           NONE   

    State StateName               County CountyState  ...  \
584    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
630    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
265    AK    ALASKA  SOUTHEAST FAIRBANKS          AK  ...   
582    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
353    AK    ALASKA    MATANUSKA-SUSITNA          AK  ...   
..    ...       ...                  ...         ...  ...   
183    AK    ALASKA      SKAGWAY-YAKUTAT          AK  ...   
516    AK    ALASKA  WRANGELL-PETERSBURG          AK  ...   
363    AK    ALASKA    KETCHIKAN GATEWAY          AK  ...   
382    AK    ALASKA    KETCHIKAN GATEWAY          AK  ...   
628    AK    ALASKA    KETCHIKAN GATEWAY          AK  ...   

    AirportPositionSource AirportPositionSourceDate AirportElevationSource  \
584                   NaN                       NaN                    NaN   
630                 OWNER                  00:00:00                  OWNER   
265            FAA OE/AAA                  00:00:00             FAA OE/AAA   
582                   NaN                       NaN                    NaN   
353                 OWNER                  00:00:00                  OWNER   
..                    ...                       ...                    ...   
183                   NaN                       NaN                    NaN   
516                   NaN                       NaN                    NaN   
363                   NaN                       NaN                    NaN   
382                   NaN                       NaN                    NaN   
628                   NaN                       NaN                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
584                        NaN                   NaN              NaN   
630                   00:00:00                   NaN              NaN   
265                   00:00:00                   NaN              NaN   
582                        NaN                   NaN              NaN   
353                   00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
183                        NaN                   NaN              NaN   
516                        NaN                   NaN              NaN   
363                        NaN                   NaN              NaN   
382                        NaN                   NaN              NaN   
628                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
584           NaN             Y            NaN             NaN  
630         CARGO             Y            NaN             NaN  
265           NaN             Y            NaN             NaN  
582           NaN           NaN            NaN             NaN  
353           NaN             Y            NaN             NaN  
..            ...           ...            ...             ...  
183           NaN           Na

#### When ascending = True, the values are arranged from smallest to largest. This means that with ascending = False, the values will be arranged from largest to smallest, allowing us to sort the aviation facilities from highest to lowest.

To sort the dataset, you need to know:

The name(s) of column(s) to sort by, and
The sorting order.
After sorting the dataset, we'll extract the first five rows which will contain the five highest aviation facilities, thus answering Question 3. This can be done with head method of Pandas dataframe.

In [12]:
df_facilities.sort_values(by = "ARPElevation", ascending = False).head(n = 5)

SiteNumber     Type LocationID EffectiveDate Region DistrictOffice State  \
584   50739.1*A  AIRPORT      'AA09     3/30/2017    AAL           NONE    AK   
630  50738.20*A  AIRPORT      '93AK     3/30/2017    AAL           NONE    AK   
265   50324.6*A  AIRPORT       '4Z5     3/30/2017    AAL           NONE    AK   
582  50738.01*A  AIRPORT      '2AK7     3/30/2017    AAL           NONE    AK   
353  50439.18*A  AIRPORT      'AK80     3/30/2017    AAL           NONE    AK   

    StateName               County CountyState  ... AirportPositionSource  \
584    ALASKA    MATANUSKA-SUSITNA          AK  ...                   NaN   
630    ALASKA    MATANUSKA-SUSITNA          AK  ...                 OWNER   
265    ALASKA  SOUTHEAST FAIRBANKS          AK  ...            FAA OE/AAA   
582    ALASKA    MATANUSKA-SUSITNA          AK  ...                   NaN   
353    ALASKA    MATANUSKA-SUSITNA          AK  ...                 OWNER   

    AirportPositionSourceDate AirportElevationSource  \
584                       NaN                    NaN   
630                  00:00:00                  OWNER   
265                  00:00:00             FAA OE/AAA   
582                       NaN                    NaN   
353                  00:00:00                  OWNER   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
584                        NaN                   NaN              NaN   
630                   00:00:00                   NaN              NaN   
265                   00:00:00                   NaN              NaN   
582                        NaN                   NaN              NaN   
353                   00:00:00                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
584           NaN             Y            NaN             NaN  
630         CARGO             Y            NaN             NaN  
265           NaN             Y            NaN             NaN  
582           NaN           NaN            NaN             NaN  
353           NaN             Y            NaN             NaN  

[5 rows x 103 columns]

#### The number of rows returned is determined by n parameter of method head. The result of the above code is the answer to Question 3. 

#### Question 4: What is the average elevation of aviation facilities?
In this dataset, we know that ARPElevation tells us the elevation of aviation facilities in Alaska. So how do we find the average or, in other words, the mean of those numbers?

Pandas has a built in function mean() that we can use to easily calculate the average of that column.

Let's take our dataframe and apply that function!

In [13]:
df_facilities["ARPElevation"].mean()

425.93209054593876

#### As we can see from that result, the average elevation of aviation facilities is almost 426 ft.

#### Question 5: What are the highest and lowest elevation values of the aviation facilities?
How do we go about solving this question? Well, similar to the previous question, Pandas has built-in functions that help us find the maximum (largest) and minimum (lowest) values in a column. These functions are max() and min(), respectively.

Given the process we took to find the average elevation in the last question, try finding the highest and lowest elevation on your own!

Task A: Find the highest elevation value 

In [18]:
print(df_facilities["ARPElevation"].max())
print(df_facilities["ARPElevation"].mean())


3984
425.93209054593876


##### Task B: Find the lowest elevation value

In [21]:
print(df_facilities["ARPElevation"].min())


0


#### Exercise 5: Write to xlsx Files!
After getting the results we want, we often want to save those results on our computer for future use. We can save a dataframe to our computer directly in Python!

For example, let's say we want to store the answer to Question 1 in a file called seaplane.xlsx.

First, let's save the result to a local variable:

In [22]:
df_seaplane = df_facilities[df_facilities["Type"] == "SEAPLANE BASE"]

#### Now, the answer to Question 1 is saved in the variable df_seaplane.

To save a dataframe in xlsx format, we can use the to_excel method. This will write the contents of the object, df_seaplane, into an Excel xlsx file.

In [28]:
df_seaplane.to_excel("C:/Users/Michael Pannwitz/Desktop/Python for Data Science/Analyzing Data/seaplane.xlsx")

In [30]:
df = pd.read_excel ('C:/Users/Michael Pannwitz/Desktop/Python for Data Science/Analyzing Data/seaplane.xlsx')

In [31]:
df

Unnamed: 0  SiteNumber           Type LocationID EffectiveDate Region  \
0             3   50017.1*C  SEAPLANE BASE       'KKI     3/30/2017    AAL   
1             5    50022.*C  SEAPLANE BASE       'KQA     3/30/2017    AAL   
2             8  50027.52*C  SEAPLANE BASE       'Z33     3/30/2017    AAL   
3            15   50031.5*C  SEAPLANE BASE      'AK81     3/30/2017    AAL   
4            24   50035.2*C  SEAPLANE BASE       '3C3     3/30/2017    AAL   
..          ...         ...            ...        ...           ...    ...   
133         725  50877.32*C  SEAPLANE BASE       'MFN     3/30/2017    AAL   
134         734  50877.01*C  SEAPLANE BASE       '2X2     3/30/2017    AAL   
135         737    50905.*C  SEAPLANE BASE       '68A     3/30/2017    AAL   
136         746  50920.12*C  SEAPLANE BASE       '2Y3     3/30/2017    AAL   
137         749    50928.*C  SEAPLANE BASE       '78K     3/30/2017    AAL   

    DistrictOffice State StateName               County  ...  \
0             NONE    AK    ALASKA               BETHEL  ...   
1             NONE    AK    ALASKA       ALEUTIANS EAST  ...   
2             NONE    AK    ALASKA          BRISTOL BAY  ...   
3             NONE    AK    ALASKA        KODIAK ISLAND  ...   
4             NONE    AK    ALASKA            ANCHORAGE  ...   
..             ...   ...       ...                  ...  ...   
133           NONE    AK    ALASKA    MATANUSKA-SUSITNA  ...   
134           NONE    AK    ALASKA    MATANUSKA-SUSITNA  ...   
135           NONE    AK    ALASKA  WRANGELL-PETERSBURG  ...   
136           NONE    AK    ALASKA      SKAGWAY-YAKUTAT  ...   
137           NONE    AK    ALASKA    KETCHIKAN GATEWAY  ...   

    AirportPositionSource AirportPositionSourceDate AirportElevationSource  \
0                     NaN                       NaN                    NaN   
1                   STATE                  00:00:00                  STATE   
2                     NaN                       NaN                    NaN   
3                     NaN                       NaN                    NaN   
4                   OWNER                  00:00:00                  OWNER   
..                    ...                       ...                    ...   
133                 OWNER                  00:00:00                  OWNER   
134               FAA-EST                  00:00:00                  STATE   
135                   NaN                       NaN                    NaN   
136               FAA-EST                  00:00:00                    NaN   
137                   NaN                       NaN                    NaN   

    AirportElevationSourceDate ContractFuelAvailable TransientStorage  \
0                          NaN                   NaN              NaN   
1                     00:00:00                   NaN              NaN   
2                          NaN                   NaN              TIE   
3                          NaN                   NaN              NaN   
4                     00:00:00                   NaN              NaN   
..                         ...                   ...              ...   
133                   00:00:00                   NaN              NaN   
134                   00:00:00                   NaN              TIE   
135                        NaN                   NaN              NaN   
136                        NaN                   NaN              TIE   
137                        NaN                   NaN              NaN   

    OtherServices WindIndicator IcaoIdentifier BeaconSchedule2  
0             NaN             N            NaN             NaN  
1             NaN             Y            NaN             NaN  
2             NaN           NaN            NaN             NaN  
3             NaN             N            NaN             NaN  
4             NaN             Y            NaN             NaN  
..            ...           ...            ...             ...  
133           NaN           Na